### PR TITLE
GH-103963: Make dis display names of args for intrinsics opcodes

### LINKED ISSUE #103963: dis: Show names of intrinsics
For `CALL_INTRINSIC_*` opcodes, `dis` currently just shows the number, which makes it hard to figure out what the code actually does:

```
              2 LOAD_CONST               0 ('T')
              4 CALL_INTRINSIC_1         7
```

It would be nice if this instead said

```
              2 LOAD_CONST               0 ('T')
              4 CALL_INTRINSIC_1         7 (INTRINSIC_TYPEVAR)
```

Possible implementation strategy:
- instrinsic names are defined in opcode.py
- `pycore_intrinsics.h` is generated from opcode.py (bonus: we no longer have to manually update MAX_INTRINSIC_1)
- `dis` learns to read the intrinsic names from opcode.py

### Linked PRs
* gh-104029
* gh-104177

### PR DESCRIPTION
**Pull request by jkchandalia**
Make the dis module display the names of the args for opcodes `CALL_INTRINSIC_*`.

Fixes #103963

* Issue: gh-103963

### CODE REVIEW

**Comment by jkchandalia** [comment]
FYI @iritkatriel @nanjekyejoannah @JelleZijlstra

File: Lib/test/test_dis.py [hunk 1]
```diff
@@ -247,6 +247,72 @@ def wrap_func_w_kwargs():
...
+dis_intrinsic_1_6 = """\
+  0        RESUME                   0
+
+  1        BUILD_LIST               0
+           LOAD_NAME                0 (a)
+           LIST_EXTEND              1
+           CALL_INTRINSIC_1         6 (INTRINSIC_LIST_TO_TUPLE)
+           RETURN_VALUE
+"""
+dis_intrinsic_2_1 = """\
```
Comment (JelleZijlstra) [inline Lib/test/test_dis.py]: This is quite a big block of code that will have to be updated when any of the generated code changes. Would it be possible to instead test only for the `CALL_INTRINSIC_2         1 (INTRINSIC_PREP_RERAISE_STAR)` line, so we don't have to hardcode so much bytecode?

File: Tools/build/generate_opcode_h.py [hunk 2]
```diff
@@ -172,6 +181,32 @@ def main(opcode_py, outfile='Include/opcode.h', internaloutfile='Include/interna
...
+        nobj.write(DEFINE.format("MAX_INTRINSIC_1", i))
+
+        nobj.write("\n\n")
+        nobj.write("/* Binary Functions: */\n")
+        for i, op in enumerate(opcode["_intrinsic_2_descs"]):
+            nobj.write(DEFINE.format(op, i))
+        nobj.write("\n")
+        nobj.write(DEFINE.format("MAX_INTRINSIC_2", i))
+
+        nobj.write(
```
Comment (JelleZijlstra) [inline Tools/build/generate_opcode_h.py]: Let's move all this to a variable `intrinsic_footer` like we do for the header.

**Review by JelleZijlstra — COMMENTED** [review]
Looks good, a few comments. You'll also have to add a NEWS entry.

File: Tools/build/generate_opcode_h.py [hunk 3]
```diff
@@ -172,6 +181,32 @@ def main(opcode_py, outfile='Include/opcode.h', internaloutfile='Include/interna
...
+            "typedef PyObject *(*instrinsic_func1)(PyThreadState* tstate, PyObject *value);"
+        )
+        nobj.write("\n")
+        nobj.write(
+            "typedef PyObject *(*instrinsic_func2)(PyThreadState* tstate, PyObject *value1, PyObject *value2);"
+        )
+        nobj.write("\n")
+        nobj.write("extern const instrinsic_func1 _PyIntrinsics_UnaryFunctions[];")
+        nobj.write("\n")
+        nobj.write("extern const instrinsic_func2 _PyIntrinsics_BinaryFunctions[];")
```
Comment (iritkatriel) [inline Tools/build/generate_opcode_h.py]:
```suggestion
        nobj.write("extern const instrinsic_func2 _PyIntrinsics_BinaryFunctions[];")
        nobj.write("\n")
```

Comment (iritkatriel) [reply Tools/build/generate_opcode_h.py]: (Added missing newline)

**Comment by jkchandalia** [comment]
@iritkatriel @JelleZijlstra I've reviewed your comments and updated the PR. Please let me know if you see anything else or if I should convert to normal PR.

File: Lib/test/test_dis.py [hunk 4]
```diff
@@ -942,6 +971,16 @@ def test_kw_names(self):
...
         self.do_disassembly_test(wrap_func_w_kwargs, dis_kw_names)
 
+    def test_intrinsic_1(self):
+        # Test that argrepr is displayed for CALL_INTRINSIC_1
+        self.do_disassembly_test("from math import *", dis_intrinsic_1_2)
+        self.do_disassembly_test("+a", dis_intrinsic_1_5)
+        self.do_disassembly_test("(*a,)", dis_intrinsic_1_6)
+
+    def test_intrinsic_2(self):
+        assert "CALL_INTRINSIC_2         1 (INTRINSIC_PREP_RERAISE_STAR)" \
```
Comment (JelleZijlstra) [inline Lib/test/test_dis.py]: Use `self.assertIn`

**Review by JelleZijlstra — APPROVED** [review]
Looks great!

**Comment by JelleZijlstra** [comment]
I'll give Irit some time to take another look, but I think this is ready to merge.

File: Include/internal/pycore_intrinsics.h [hunk 5]
```diff
@@ -1,26 +1,24 @@
...
-#define INTRINSIC_ASYNC_GEN_WRAP 4
-#define INTRINSIC_UNARY_POSITIVE 5
-#define INTRINSIC_LIST_TO_TUPLE 6
-
-#define MAX_INTRINSIC_1 6
+#define MAX_INTRINSIC_1                          6
 
 
 /* Binary Functions: */
+#define INTRINSIC_INVALID                        0
```
Comment (iritkatriel) [inline Include/internal/pycore_intrinsics.h]: Looks like INTRINSIC_INVALID is defined twice?

Comment (jkchandalia) [reply Include/internal/pycore_intrinsics.h]: Updated to differentiate between invalid for the two `CALL_INTRINSIC_*` opcodes.

**Review by iritkatriel — APPROVED** [review]

**Pull request merged**